File: 백준/Gold/12913. 매직 포션/매직 포션.py
# 매직 포션
import sys
input = sys.stdin.readline
import heapq as hq

N, K = map(int,input().split())
graph = []
que = []
visited = [float('inf')] * N

for i in range(N):
    row = list(input().rstrip())
    graph.append(list(map(int, row)))

for i in range(1, N):
    hq.heappush(que, (graph[0][i], i, K))
    visited[i] = graph[0][i]
    
    if K > 0:
        hq.heappush(que, (graph[0][i] / 2, i, K-1))
        visited[i] = graph[0][i] / 2

while que:
    dist, cur, potion = hq.heappop(que)
   
    for nex in range(N):
        if nex == cur:
            continue

        if visited[nex] > dist + graph[cur][nex]:
            hq.heappush(que, (dist+graph[cur][nex], nex, potion))
            visited[nex] = dist+graph[cur][nex]

        if potion and visited[nex] > dist + graph[cur][nex] / 2:
            hq.heappush(que, (dist+graph[cur][nex]/2, nex, potion-1))
            visited[nex] = dist+graph[cur][nex]/2

print(visited[1])

File: 백준/Gold/12913. 매직 포션/매직 포션.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        double[][] graph = new double[N][N];
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(a[0], b[0]));
        double[] visited = new double[N];
        Arrays.fill(visited, 1000001);

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            String row = st.nextToken();
            for (int j = 0; j < N; j++) {
                graph[i][j] = Long.parseLong(row.charAt(j) + "");
            }
        }

        for (int i = 1; i < N; i++) {
            pq.offer(new double[] {graph[0][i], i , K});
            visited[i] = graph[0][i];

            if (K > 0) {
                pq.offer(new double[] {graph[0][i] / 2, i, K-1});
                visited[i] = graph[0][i] / 2;
            }
        }

        while (!pq.isEmpty()) {
            double[] tmp = pq.poll();

            double dist = tmp[0];
            int cur = (int) tmp[1];
            double potion = tmp[2];

            for (int next = 1; next < N; next++) {
                if (next == cur) {
                    continue;
                }

                if (visited[next] > dist + graph[cur][next]) {
                    pq.offer(new double[] {dist + graph[cur][next], next, potion});
                    visited[next] = dist + graph[cur][next];
                }

                if (visited[next] > dist + graph[cur][next] / 2 && potion > 0) {
                    pq.offer(new double[] {dist + graph[cur][next] / 2, next, potion-1});
                    visited[next] = dist + graph[cur][next] / 2;
                }
            }
        }
        System.out.println(visited[1]);
    }
}
